// nepse-api-unofficial/dist/index.js
// NEPSE Unofficial API - https://github.com/surajrimal07/nepse-api-unofficial

// src/logger.ts
var logger = {
  info(message, meta) {
    meta ? console.log("[INFO]", message, meta) : console.log("[INFO]", message);
  },
  warn(message, meta) {
    meta ? console.warn("[WARN]", message, meta) : console.warn("[WARN]", message);
  },
  error(message, meta) {
    meta ? console.error("[ERROR]", message, meta) : console.error("[ERROR]", message);
  }
};

// src/worker/workerPool.ts
var defaultLogger = {
  log: () => {
  },
  warn: () => {
  },
  error: () => {
  }
};
var WORKER_ID_REGEX = /^worker-(\d+)$/;
var WorkerPool = class {
  workers = /* @__PURE__ */ new Map();
  taskQueue = [];
  pendingTasks = /* @__PURE__ */ new Map();
  logger;
  // Configuration
  minWorkers;
  maxWorkers;
  idleTimeout;
  taskTimeout;
  maxRetries;
  scalingThreshold;
  // State tracking
  taskIdCounter = 0;
  workerIdCounter = 0;
  availableWorkerIds = /* @__PURE__ */ new Set();
  // Track available worker IDs for reuse
  cleanupInterval = null;
  isShuttingDown = false;
  constructor(options = {}) {
    this.logger = options.logger ?? defaultLogger;
    this.minWorkers = options.minWorkers ?? 1;
    this.maxWorkers = options.maxWorkers ?? 10;
    this.idleTimeout = options.idleTimeout ?? 3e4;
    this.taskTimeout = options.taskTimeout ?? 6e4;
    this.maxRetries = options.maxRetries ?? 3;
    this.scalingThreshold = options.scalingThreshold ?? 0.6;
    this.initializeWorkers();
    this.startCleanupInterval();
  }
  initializeWorkers() {
    for (let i = 0; i < this.minWorkers; i++) {
      this.spawnWorker();
    }
  }
  generateTaskId() {
    return `task-${++this.taskIdCounter}`;
  }
  generateWorkerId() {
    let workerId;
    if (this.availableWorkerIds.size > 0) {
      workerId = Math.min(...this.availableWorkerIds);
      this.availableWorkerIds.delete(workerId);
    } else {
      workerId = ++this.workerIdCounter;
    }
    return `worker-${workerId}`;
  }
  spawnWorker() {
    const workerId = this.generateWorkerId();
    const worker = new Worker(
      new URL("./worker.js", import.meta.url).href
    );
    const wrapper = {
      id: workerId,
      worker,
      busy: false,
      lastUsed: Date.now(),
      taskCount: 0,
      errorCount: 0,
      activeTask: null
    };
    worker.onmessage = (event) => {
      this.handleWorkerMessage(workerId, event.data);
    };
    worker.onerror = (error) => {
      this.logger.error(`Worker ${workerId} error:`, error);
      this.handleWorkerError(workerId);
    };
    worker.addEventListener("close", () => {
      this.logger.log(`Worker ${workerId} closed`);
      this.workers.delete(workerId);
    });
    this.workers.set(workerId, wrapper);
    this.logger.log(
      `Spawned worker ${workerId}. Total workers: ${this.workers.size}`
    );
    return wrapper;
  }
  handleWorkerMessage(workerId, message) {
    const worker = this.workers.get(workerId);
    if (!worker) {
      return;
    }
    const pendingTask = this.pendingTasks.get(message.taskId);
    if (!pendingTask) {
      this.logger.warn(`Received response for unknown task: ${message.taskId}`);
      return;
    }
    if (pendingTask.timeoutId) {
      clearTimeout(pendingTask.timeoutId);
    }
    worker.busy = false;
    worker.lastUsed = Date.now();
    worker.activeTask = null;
    worker.taskCount++;
    this.pendingTasks.delete(message.taskId);
    if (message.success && message.error === void 0) {
      pendingTask.resolve(message.result);
    } else {
      worker.errorCount++;
      const error = new Error(message.error || "Worker task failed");
      if (pendingTask.retryCount < this.maxRetries) {
        this.logger.warn(
          `Retrying task ${message.taskId}, attempt ${pendingTask.retryCount + 1}`
        );
        pendingTask.retryCount++;
        this.taskQueue.unshift(pendingTask);
        this.processQueue();
      } else {
        pendingTask.reject(error);
      }
    }
    this.processQueue();
  }
  handleWorkerError(workerId) {
    const worker = this.workers.get(workerId);
    if (!worker) {
      return;
    }
    worker.errorCount++;
    if (worker.activeTask) {
      const pendingTask = this.pendingTasks.get(worker.activeTask);
      if (pendingTask && pendingTask.retryCount < this.maxRetries) {
        this.logger.warn(`Worker error, retrying task ${worker.activeTask}`);
        pendingTask.retryCount++;
        this.taskQueue.unshift(pendingTask);
        this.pendingTasks.delete(worker.activeTask);
      }
    }
    worker.busy = false;
    worker.activeTask = null;
    if (worker.errorCount > 5) {
      this.logger.warn(
        `Terminating worker ${workerId} due to excessive errors`
      );
      this.terminateWorker(workerId);
    }
    this.processQueue();
  }
  getAvailableWorker() {
    for (const worker of this.workers.values()) {
      if (!(worker.busy || this.isShuttingDown)) {
        return worker;
      }
    }
    return null;
  }
  shouldScaleUp() {
    if (this.workers.size >= this.maxWorkers) {
      return false;
    }
    const busyWorkers = Array.from(this.workers.values()).filter(
      (w) => w.busy
    ).length;
    const busyRatio = busyWorkers / this.workers.size;
    return busyRatio >= this.scalingThreshold && this.taskQueue.length > 0;
  }
  processQueue() {
    if (this.isShuttingDown || this.taskQueue.length === 0) {
      return;
    }
    while (this.taskQueue.length > 0) {
      let availableWorker = this.getAvailableWorker();
      if (!availableWorker && this.shouldScaleUp()) {
        this.logger.log(
          `Scaling up: spawning new worker. Queue length: ${this.taskQueue.length}`
        );
        availableWorker = this.spawnWorker();
      }
      if (!availableWorker) {
        break;
      }
      const task = this.taskQueue.shift();
      if (!task) {
        break;
      }
      this.assignTaskToWorker(availableWorker, task);
    }
  }
  assignTaskToWorker(worker, task) {
    worker.busy = true;
    worker.lastUsed = Date.now();
    worker.activeTask = task.taskId;
    const timeoutId = setTimeout(() => {
      this.handleTaskTimeout(task.taskId);
    }, this.taskTimeout);
    task.timeoutId = timeoutId;
    const workerTask = {
      taskId: task.taskId,
      action: task.action,
      payload: task.payload,
      timestamp: Date.now()
    };
    worker.worker.postMessage(workerTask);
    this.logger.log(`Assigned task ${task.taskId} to worker ${worker.id}`);
  }
  handleTaskTimeout(taskId) {
    const task = this.pendingTasks.get(taskId);
    if (!task) {
      return;
    }
    this.logger.warn(`Task ${taskId} timed out`);
    for (const worker of this.workers.values()) {
      if (worker.activeTask === taskId) {
        worker.busy = false;
        worker.activeTask = null;
        worker.errorCount++;
        break;
      }
    }
    this.pendingTasks.delete(taskId);
    if (task.retryCount < this.maxRetries) {
      task.retryCount++;
      const oldTaskId = task.taskId;
      const newTaskId = this.generateTaskId();
      task.taskId = newTaskId;
      this.logger.log(
        `Retrying task ${oldTaskId} as ${newTaskId}, attempt ${task.retryCount}`
      );
      this.pendingTasks.set(newTaskId, task);
      this.taskQueue.unshift(task);
      this.processQueue();
    } else {
      task.reject(
        new Error(`Task ${taskId} timed out after ${this.maxRetries} retries`)
      );
    }
  }
  startCleanupInterval() {
    this.cleanupInterval = setInterval(() => {
      this.cleanupIdleWorkers();
      this.cleanupStaleTask();
    }, 1e4);
  }
  cleanupIdleWorkers() {
    if (this.workers.size <= this.minWorkers) {
      return;
    }
    const now = Date.now();
    const workersToRemove = [];
    for (const [workerId, worker] of this.workers) {
      if (!worker.busy && now - worker.lastUsed > this.idleTimeout) {
        workersToRemove.push(workerId);
      }
    }
    const canRemove = Math.min(
      workersToRemove.length,
      this.workers.size - this.minWorkers
    );
    for (let i = 0; i < canRemove; i++) {
      this.terminateWorker(workersToRemove[i]);
    }
  }
  cleanupStaleTask() {
    const now = Date.now();
    const staleTasks = [];
    for (const [taskId, task] of this.pendingTasks) {
      if (now - task.timestamp > this.taskTimeout * 2) {
        staleTasks.push(taskId);
      }
    }
    for (const taskId of staleTasks) {
      const task = this.pendingTasks.get(taskId);
      if (task) {
        this.pendingTasks.delete(taskId);
        task.reject(new Error(`Task ${taskId} became stale`));
      }
    }
  }
  terminateWorker(workerId) {
    const worker = this.workers.get(workerId);
    if (!worker) {
      return;
    }
    this.logger.log(`Terminating worker ${workerId}`);
    const workerMatch = workerId.match(WORKER_ID_REGEX);
    if (workerMatch) {
      const workerNumber = Number.parseInt(workerMatch[1], 10);
      this.availableWorkerIds.add(workerNumber);
    }
    if (worker.activeTask) {
      const task = this.pendingTasks.get(worker.activeTask);
      if (task && task.retryCount < this.maxRetries) {
        task.retryCount++;
        this.logger.log(
          `Retrying task ${task.taskId} due to worker termination, attempt ${task.retryCount}`
        );
        const newTaskId = this.generateTaskId();
        task.taskId = newTaskId;
        this.pendingTasks.set(newTaskId, task);
        this.taskQueue.unshift(task);
      }
      this.pendingTasks.delete(worker.activeTask);
    }
    worker.worker.terminate();
    this.workers.delete(workerId);
    this.logger.log(
      `Worker ${workerId} terminated. Remaining workers: ${this.workers.size}`
    );
  }
  // Public API
  runTask(action, payload) {
    if (this.isShuttingDown) {
      throw new Error("Worker pool is shutting down");
    }
    return new Promise((resolve, reject) => {
      const taskId = this.generateTaskId();
      const task = {
        taskId,
        resolve,
        reject,
        timestamp: Date.now(),
        retryCount: 0,
        action,
        payload
      };
      this.pendingTasks.set(taskId, task);
      this.taskQueue.push(task);
      this.processQueue();
    });
  }
  getStats() {
    const workers = Array.from(this.workers.values());
    return {
      totalWorkers: workers.length,
      busyWorkers: workers.filter((w) => w.busy).length,
      idleWorkers: workers.filter((w) => !w.busy).length,
      queueLength: this.taskQueue.length,
      pendingTasks: this.pendingTasks.size,
      totalTasksCompleted: workers.reduce((sum, w) => sum + w.taskCount, 0),
      totalErrors: workers.reduce((sum, w) => sum + w.errorCount, 0)
    };
  }
  async shutdown() {
    this.logger.log("Shutting down worker pool...");
    this.isShuttingDown = true;
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
    }
    for (const task of this.pendingTasks.values()) {
      task.reject(new Error("Worker pool is shutting down"));
    }
    this.pendingTasks.clear();
    this.taskQueue.length = 0;
    const terminationPromises = Array.from(this.workers.keys()).map(
      (workerId) => {
        return new Promise((resolve) => {
          const worker = this.workers.get(workerId);
          if (worker) {
            worker.worker.addEventListener("close", () => resolve());
            worker.worker.terminate();
          } else {
            resolve();
          }
        });
      }
    );
    await Promise.all(terminationPromises);
    this.workers.clear();
    this.logger.log("Worker pool shutdown complete");
  }
};

// src/worker/index.ts
var workerPoolInstance = null;
function getWorkerPool() {
  if (!workerPoolInstance) {
    const defaultOptions = {
      minWorkers: 1,
      maxWorkers: 10,
      idleTimeout: 3e4,
      // 30 seconds
      taskTimeout: 6e4,
      // 60 seconds
      maxRetries: 3,
      scalingThreshold: 0.8,
      // Scale when 80% of workers are busy
      logger: {
        log: (message) => logger.info(`[WorkerPool] ${message}`),
        warn: (message) => logger.warn(`[WorkerPool] ${message}`),
        error: (message, error) => logger.error(`[WorkerPool] ${message}`, error)
      }
    };
    workerPoolInstance = new WorkerPool(defaultOptions);
  }
  return workerPoolInstance;
}
async function shutdownWorkerPool() {
  if (workerPoolInstance) {
    await workerPoolInstance.shutdown();
    workerPoolInstance = null;
  }
}
function getWorkerPoolStats() {
  return workerPoolInstance?.getStats() || null;
}
async function get_market_status() {
  try {
    return await getWorkerPool().runTask(
      "get_market_status"
    );
  } catch {
    return null;
  }
}
async function getSummary() {
  try {
    return await getWorkerPool().runTask("getSummary");
  } catch {
    return null;
  }
}
async function getTopTenTradeScrips() {
  try {
    return await getWorkerPool().runTask(
      "getTopTenTradeScrips"
    );
  } catch {
    return null;
  }
}
async function getTopTenTransactions() {
  try {
    return await getWorkerPool().runTask(
      "getTopTenTransactions"
    );
  } catch {
    return null;
  }
}
async function getTopTenTurnover() {
  try {
    return await getWorkerPool().runTask(
      "getTopTenTurnover"
    );
  } catch {
    return null;
  }
}
async function getTopGainers() {
  try {
    return await getWorkerPool().runTask(
      "getTopGainers"
    );
  } catch {
    return null;
  }
}
async function getTopLoosers() {
  try {
    return await getWorkerPool().runTask(
      "getTopLoosers"
    );
  } catch {
    return null;
  }
}
async function getNepseIndex() {
  try {
    return await getWorkerPool().runTask(
      "getNepseIndex"
    );
  } catch {
    return null;
  }
}
async function getNepseIndexIntraday(index = "NEPSE Index") {
  try {
    return await getWorkerPool().runTask("getNepseIndexIntraday", {
      index
    });
  } catch {
    return null;
  }
}
async function stockIntraday(stock = "NABIL") {
  try {
    return await getWorkerPool().runTask("stockIntraday", {
      stock
    });
  } catch {
    return null;
  }
}
async function getPriceVolume() {
  try {
    return await getWorkerPool().runTask(
      "getPriceVolume"
    );
  } catch {
    return null;
  }
}
async function getSecurityList() {
  try {
    return await getWorkerPool().runTask(
      "getSecurityList"
    );
  } catch {
    return null;
  }
}
async function getCompaniesList() {
  try {
    return await getWorkerPool().runTask(
      "getCompaniesList"
    );
  } catch {
    return null;
  }
}
async function getMarket_depth(symbol) {
  try {
    return await getWorkerPool().runTask("getMarket_depth", {
      symbol
    });
  } catch {
    return null;
  }
}
async function getSupplyDemand() {
  try {
    return await getWorkerPool().runTask(
      "getSupplyDemand"
    );
  } catch {
    return null;
  }
}
async function getPriceVolumeHistory(symbol, size = 500) {
  try {
    return await getWorkerPool().runTask("getPriceVolumeHistory", {
      symbol,
      size
    });
  } catch {
    return null;
  }
}
async function getIndexPriceVolumeHistory(index, size = 500) {
  try {
    return await getWorkerPool().runTask("getIndexPriceVolumeHistory", {
      index,
      size
    });
  } catch {
    return null;
  }
}
async function getStockDailyPrice(symbol) {
  try {
    return await getWorkerPool().runTask("getStockDailyPrice", {
      symbol
    });
  } catch {
    return null;
  }
}
async function liveMarketData() {
  try {
    return await getWorkerPool().runTask(
      "liveMarketData"
    );
  } catch {
    return null;
  }
}
async function disclosure() {
  try {
    return await getWorkerPool().runTask(
      "disclosure"
    );
  } catch {
    return null;
  }
}
async function get_security_detail(symbol) {
  try {
    return await getWorkerPool().runTask("get_security_detail", {
      symbol
    });
  } catch {
    return null;
  }
}

// src/API_ENDPOINTS.json
var API_ENDPOINTS_default = {
  nepse_open_url: "/api/nots/nepse-data/market-open",
  security_list_url: "/api/nots/security?nonDelisted=true",
  wasm_url: "/assets/prod/css.wasm",
  prove_url: "/api/authenticate/prove",
  refresh_url: "/api/authenticate/refresh-token"};

// src/constants.ts
var BASE_URL = "https://www.nepalstock.com.np";

// src/error.ts
function createNepseError(message) {
  const error = new Error(message);
  error.name = "NepseError";
  return error;
}

// src/headers.ts
function createNepseHeaders(token) {
  return {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36",
    Accept: "application/json, text/plain, */*",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-US,en;q=0.9",
    Connection: "keep-alive",
    Referer: "www.nepalstock.com.np",
    Authorization: `Salter ${token}`,
    "Content-Type": "application/json",
    Host: "www.nepalstock.com.np"
  };
}

// src/token-manager.ts
var TokenManager = class _TokenManager {
  static instance;
  static wasm_instantiated = false;
  accessToken = null;
  refreshToken = null;
  tokenTimestamp = null;
  proveObject = null;
  marketStatus = null;
  MAX_TOKEN_AGE = 45e3;
  // 45 seconds in milliseconds
  isUpdating = false;
  updatePromise = null;
  security_brief_cache = [];
  security_symbol_id_keymap = null;
  // WASM related properties
  static cdx = null;
  static rdx = null;
  static bdx = null;
  static ndx = null;
  static mdx = null;
  constructor() {
  }
  //call this method before using the token manager
  static async initializeWasm() {
    if (_TokenManager.wasm_instantiated) {
      return;
    }
    const response = await fetch(`${BASE_URL}${API_ENDPOINTS_default.wasm_url}`, {
      tls: { rejectUnauthorized: false }
    });
    if (!response.ok) {
      throw createNepseError("Failed to fetch WASM file");
    }
    const wasm = await response.arrayBuffer().then((buffer) => {
      return WebAssembly.instantiate(buffer);
    });
    _TokenManager.cdx = wasm.instance.exports.cdx;
    _TokenManager.rdx = wasm.instance.exports.rdx;
    _TokenManager.bdx = wasm.instance.exports.bdx;
    _TokenManager.ndx = wasm.instance.exports.ndx;
    _TokenManager.mdx = wasm.instance.exports.mdx;
    _TokenManager.wasm_instantiated = true;
  }
  static async getInstance() {
    if (!_TokenManager.wasm_instantiated) {
      await _TokenManager.initializeWasm();
    }
    if (!_TokenManager.instance) {
      _TokenManager.instance = new _TokenManager();
    }
    return _TokenManager.instance;
  }
  async updateToken() {
    try {
      if (this.refreshToken && !this.isTokenValid()) {
        await this.refreshAccessToken();
      } else {
        this.proveObject = await this.get_raw_access_object();
        if (!this.proveObject) {
          throw createNepseError("Prove object is null or undefined");
        }
        this.accessToken = this.get_valid_token();
        this.tokenTimestamp = Date.now();
        this.marketStatus = await this.fetchMarketStatus();
      }
    } catch (error) {
      this.accessToken = null;
      this.tokenTimestamp = null;
      this.marketStatus = null;
      throw error;
    } finally {
      this.isUpdating = false;
      this.updatePromise = null;
    }
  }
  async refreshAccessToken() {
    const response = await fetch(`${BASE_URL}${API_ENDPOINTS_default.refresh_url}`, {
      method: "POST",
      // biome-ignore lint/style/noNonNullAssertion: <ikmow>
      headers: createNepseHeaders(this.accessToken),
      body: JSON.stringify({ refreshToken: this.refreshToken })
    });
    if (!response.ok) {
      throw createNepseError("Failed to refresh access token");
    }
    const data = await response.json();
    this.accessToken = data.accessToken;
    this.refreshToken = data.refreshToken;
    this.tokenTimestamp = Date.now();
  }
  get_valid_token() {
    if (!_TokenManager.wasm_instantiated) {
      throw createNepseError(
        "please call instantiate method before doing anything"
      );
    }
    if (!this.proveObject) {
      throw createNepseError("Prove object is not initialized");
    }
    const { accessToken, salt1, salt2, salt3, salt4, salt5 } = this.proveObject;
    const cdx = _TokenManager.cdx?.(salt1, salt2, salt3, salt4, salt5);
    const rdx = _TokenManager.rdx?.(salt1, salt2, salt4, salt3, salt5);
    const bdx = _TokenManager.bdx?.(salt1, salt2, salt4, salt3, salt5);
    const ndx = _TokenManager.ndx?.(salt1, salt2, salt4, salt3, salt5);
    const mdx = _TokenManager.mdx?.(salt1, salt2, salt4, salt3, salt5);
    return (accessToken.slice(0, cdx) + accessToken.slice(cdx + 1, rdx) + accessToken.slice(rdx + 1, bdx) + accessToken.slice(bdx + 1, ndx) + accessToken.slice(ndx + 1, mdx) + accessToken.slice(mdx + 1)).trim();
  }
  get_valid_body_id() {
    const dummyData = [
      147,
      117,
      239,
      143,
      157,
      312,
      161,
      612,
      512,
      804,
      411,
      527,
      170,
      511,
      421,
      667,
      764,
      621,
      301,
      106,
      133,
      793,
      411,
      511,
      312,
      423,
      344,
      346,
      653,
      758,
      342,
      222,
      236,
      811,
      711,
      611,
      122,
      447,
      128,
      199,
      183,
      135,
      489,
      703,
      800,
      745,
      152,
      863,
      134,
      211,
      142,
      564,
      375,
      793,
      212,
      153,
      138,
      153,
      648,
      611,
      151,
      649,
      318,
      143,
      117,
      756,
      119,
      141,
      717,
      113,
      112,
      146,
      162,
      660,
      693,
      261,
      362,
      354,
      251,
      641,
      157,
      178,
      631,
      192,
      734,
      445,
      192,
      883,
      187,
      122,
      591,
      731,
      852,
      384,
      565,
      596,
      451,
      772,
      624,
      691
    ];
    const currentDate = /* @__PURE__ */ new Date();
    const datePart = currentDate.getDate();
    if (!this.marketStatus) {
      throw createNepseError("Market id is not set");
    }
    const id = dummyData[this.marketStatus.id] + this.marketStatus.id + 2 * datePart;
    return id;
  }
  getPOSTPayloadID() {
    if (!this.proveObject) {
      throw createNepseError("Prove object or market status not available");
    }
    const e = this.get_valid_body_id();
    const currentDay = (/* @__PURE__ */ new Date()).getDate();
    const saltIndex = e % 10 < 5 ? 3 : 1;
    const salts = [
      this.proveObject.salt1,
      this.proveObject.salt2,
      this.proveObject.salt3,
      this.proveObject.salt4,
      this.proveObject.salt5
    ];
    const finalId = e + salts[saltIndex] * currentDay - salts[saltIndex - 1];
    return finalId;
  }
  getPOSTPayloadIDForFloorSheet() {
    if (!this.proveObject) {
      throw createNepseError("Prove object or market status not available");
    }
    const e = this.get_valid_body_id();
    const currentDay = (/* @__PURE__ */ new Date()).getDate();
    const saltIndex = e % 10 < 4 ? 1 : 3;
    const salts = [
      this.proveObject.salt1,
      this.proveObject.salt2,
      this.proveObject.salt3,
      this.proveObject.salt4,
      this.proveObject.salt5
    ];
    const finalId = e + salts[saltIndex] * currentDay - salts[saltIndex - 1];
    return finalId;
  }
  async get_raw_access_object() {
    const proveResponse = await fetch(
      `${BASE_URL}${API_ENDPOINTS_default.prove_url}`,
      {
        headers: {
          "User-Agent": "Mozilla/5.0",
          Referer: `${BASE_URL}`
        },
        method: "GET"
      }
    ).then((response) => {
      return response.json();
    }).then((data) => data).catch(() => {
      return null;
    });
    return proveResponse;
  }
  async fetchMarketStatus() {
    const response = await fetch(`${BASE_URL}${API_ENDPOINTS_default.nepse_open_url}`, {
      // biome-ignore lint/style/noNonNullAssertion: <iknow>
      headers: createNepseHeaders(this.accessToken),
      method: "GET"
    });
    if (!response.ok) {
      throw createNepseError("Failed to fetch market status");
    }
    return response.json();
  }
  isTokenValid() {
    if (!(this.tokenTimestamp && this.accessToken)) {
      return false;
    }
    return Date.now() - this.tokenTimestamp < this.MAX_TOKEN_AGE;
  }
  async getAccessToken() {
    if (this.isTokenValid()) {
      return this.accessToken;
    }
    if (!this.isUpdating) {
      this.isUpdating = true;
      this.updatePromise = this.updateToken();
    }
    await this.updatePromise;
    if (!this.accessToken) {
      throw createNepseError("Failed to get access token");
    }
    return this.accessToken;
  }
  getProveObject() {
    return this.proveObject;
  }
  async get_security_briefs() {
    const token = await this.getAccessToken();
    const securityBriefDetails = await fetch(
      `${BASE_URL}${API_ENDPOINTS_default.security_list_url}`,
      {
        headers: {
          "User-Agent": "Mozilla/5.0",
          Referer: `${BASE_URL}`,
          Authorization: `Salter ${token}`
        },
        method: "GET"
      }
    ).then((response) => response.json()).then((data) => data).catch((err) => {
      console.error(err);
      return [];
    });
    return securityBriefDetails;
  }
  getSalts() {
    if (!this.proveObject) {
      return null;
    }
    return [
      this.proveObject.salt1,
      this.proveObject.salt2,
      this.proveObject.salt3,
      this.proveObject.salt4,
      this.proveObject.salt5
    ];
  }
  getMarketId() {
    return this.marketStatus?.id ?? null;
  }
  async getSecurityBrief() {
    if (this.security_brief_cache.length === 0) {
      this.security_brief_cache = await this.get_security_briefs();
    }
    return this.security_brief_cache;
  }
  async getSecuritySymbolIdMap() {
    if (this.security_symbol_id_keymap === null) {
      const securities = await this.getSecurityBrief();
      this.security_symbol_id_keymap = securities.reduce(
        (map, security) => {
          map[security.symbol] = security.id;
          return map;
        },
        {}
      );
    }
    return this.security_symbol_id_keymap;
  }
  getMarketStatus() {
    return this.marketStatus;
  }
};
var tokenManager = TokenManager.getInstance();

// src/types/indexes.ts
var nepseIndexes = [
  "Banking SubIndex",
  "Development Bank Ind.",
  "Finance Index",
  "Hotels And Tourism",
  "HydroPower Index",
  "Investment",
  "Life Insurance",
  "Manufacturing And Pr.",
  "Microfinance Index",
  "Mutual Fund",
  "NEPSE Index",
  "Non Life Insurance",
  "Others Index",
  "Trading Index"
];

// src/types/types.ts
var INDEX_URL_MAP = {
  "Banking SubIndex": "DailyBankSubindexGraph",
  "Development Bank Ind.": "DailyDevelopmentBankSubindexGraph",
  "Finance Index": "DailyFinanceSubindexGraph",
  "Hotels And Tourism": "DailyHotelTourismSubindexGraph",
  "HydroPower Index": "DailyHydroPowerSubindexGraph",
  Investment: "DailyInvestmentSubindexGraph",
  "Life Insurance": "DailyLifeInsuranceSubindexGraph",
  "Manufacturing And Pr.": "DailyManufacturingProcessingSubindexGraph",
  "Microfinance Index": "DailyMicrofinanceSubindexGraph",
  "Mutual Fund": "DailyMutualFundSubindexGraph",
  "NEPSE Index": "DailyNepseIndexGraph",
  "Non Life Insurance": "DailyNonLifeInsuranceSubindexGraph",
  "Others Index": "DailyOthersSubindexGraph",
  "Trading Index": "DailyTradingSubindexGraph"
};
var NepseSources = /* @__PURE__ */ ((NepseSources2) => {
  NepseSources2["AbhiyanDaily"] = "Abhiyan Daily";
  NepseSources2["Arthasarokar"] = "Arthasarokar";
  NepseSources2["KarobarDaily"] = "Karobar Daily";
  NepseSources2["Bizmandu"] = "Bizmandu";
  NepseSources2["Arthapath"] = "Arthapath";
  NepseSources2["Clickmandu"] = "Clickmandu";
  NepseSources2["CapitalNepal"] = "Capital Nepal";
  NepseSources2["BizKhabar"] = "BizKhabar";
  NepseSources2["ArthaKendraNepali"] = "ArthaKendra Nepali";
  NepseSources2["ArthaKendraEnglish"] = "ArthaKendra English";
  NepseSources2["ShareSansar"] = "Share Sansar";
  NepseSources2["MeroLagani"] = "Mero Lagani";
  return NepseSources2;
})(NepseSources || {});

export { BASE_URL, INDEX_URL_MAP, NepseSources, TokenManager, WorkerPool, createNepseError, createNepseHeaders, disclosure, getCompaniesList, getIndexPriceVolumeHistory, getMarket_depth, getNepseIndex, getNepseIndexIntraday, getPriceVolume, getPriceVolumeHistory, getSecurityList, getStockDailyPrice, getSummary, getSupplyDemand, getTopGainers, getTopLoosers, getTopTenTradeScrips, getTopTenTransactions, getTopTenTurnover, getWorkerPool, getWorkerPoolStats, get_market_status, get_security_detail, liveMarketData, logger, nepseIndexes, shutdownWorkerPool, stockIntraday, tokenManager };
